backend-worker/worker.py

```diff
@@ -1,26 +1,546 @@
+            -10965275.57, 3429693.30)
+        parcel = point_over_san_antonio.buffer(100)
+
+        target_raster_path = os.path.join(self.workspace_dir, 'raster.tif')
+        fill_parcel(parcel.wkt, 15, target_raster_path)
+
+        result_array = pygeoprocessing.raster_to_numpy_array(
+                target_raster_path)
+        self.assertEqual(
+            numpy.sum(result_array[result_array != NLUD_NODATA]), 600)
+        self.assertEqual(numpy.sum(result_array == 15), 40)
+
+    def test_wallpaper(self):
+        # University of Texas: San Antonio, selected by hand in QGIS
+        # Coordinates are in EPSG:3857 "Web Mercator"
+        point_over_san_antonio = shapely.geometry.Point(
+            -10965275.57, 3429693.30)
+        parcel = point_over_san_antonio.buffer(100)
+
+        # Apache Creek, urban residential area, San Antonio, TX.
+        # Selected by hand in QGIS.  Coordinates are in EPSG:3857 "Web
+        # Mercator"
+        pattern = shapely.geometry.box(
+            *shapely.geometry.Point(
+                -10968418.16, 3429347.98).buffer(100).bounds)
+
+        target_raster_path = os.path.join(
+            self.workspace_dir, 'wallpapered_raster.tif')
+
+        nlud_path = 'appdata/nlud.tif'
+        wallpaper_parcel(parcel.wkt, pattern.wkt, nlud_path,
+                         target_raster_path, self.workspace_dir)
+
+        # This is useful for debugging
+        # import pdb; pdb.set_trace()  # print(self.workspace_dir)
+
+
+def _reproject_to_nlud(parcel_wkt_epsg3857):
+    ogr_geom = ogr.CreateGeometryFromWkt(parcel_wkt_epsg3857)
+    err_code = ogr_geom.Transform(WEB_MERCATOR_TO_ALBERS_EQ_AREA)
+    if err_code:
+        LOGGER.warning(
+            "Transformation failed on parcel; continuing with "
+            f"{parcel_wkt_epsg3857}")
+    assert ogr_geom.ExportToWkt() != parcel_wkt_epsg3857
+    parcel_geom = shapely.wkt.loads(ogr_geom.ExportToWkt())
+    return parcel_geom
+
+
+def _create_new_lulc(parcel_wkt_epsg3857, target_local_gtiff_path):
+    """Create an LULC raster in the NLUD projection covering the parcel.
+
+    Args:
+        parcel_wkt_epsg3857 (str): The parcel WKT in EPSG:3857 (Web Mercator)
+        target_local_gtiff_path (str): Where the target raster should be saved
+
+    Returns:
+        ``None``
+    """
+    parcel_geom = _reproject_to_nlud(parcel_wkt_epsg3857)
+    parcel_min_x, parcel_min_y, parcel_max_x, parcel_max_y = parcel_geom.bounds
+    buffered_parcel_geom = parcel_geom.buffer(
+        abs(min(parcel_max_x - parcel_min_x,
+                parcel_max_y - parcel_min_y)))
+    buf_minx, buf_miny, buf_maxx, buf_maxy = buffered_parcel_geom.bounds
+
+    buf_minx -= abs((buf_minx - NLUD_ORIGIN_X) % PIXELSIZE_X)
+    buf_miny -= abs((buf_miny - NLUD_ORIGIN_Y) % PIXELSIZE_Y)
+    buf_maxx += PIXELSIZE_X - abs((buf_maxx - NLUD_ORIGIN_X) % PIXELSIZE_X)
+    buf_maxy += PIXELSIZE_Y - abs((buf_maxy - NLUD_ORIGIN_Y) % PIXELSIZE_Y)
```

Comment:
Sorry, it's probably because it's late in the day, but I'm a little confused about how we're buffering here... Would you mind walking me through it or adding some comments? Thanks!

Reply:
Sure!  So the coordinates of the buffered parcel bounding box are in coordinate units and not necessarily aligned with the actual pixel boundaries.  This calculates the bounding box in terms of all of the pixels touched by the buffered parcel geometry _and is aligned_ with the NLUD's pixel grid.

So it's sort of like rounding up/down, but to pixel grid coordinates in order to get the number of rows and columns and the origin of the resulting raster.